### PR TITLE
Feat: API: Implement StateLookupRobustAddress

### PR DESCRIPTION
**Pull request by geoff-vball**
## Related Issues
Fixes https://github.com/filecoin-project/lotus/issues/5371

## Proposed Changes
added endpoint StateLoockupRobustAddress which will return the robust address corresponding to any ID Address

## Checklist

Before you mark the PR ready for review, please make sure that:
- [ ] All commits have a clear commit message.
- [ ] The PR title is in the form of of `<PR type>: <area>: <change being made>`
    - example: ` fix: mempool: Introduce a cache for valid signatures`
    - `PR type`: _fix_, _feat_, _INTERFACE BREAKING CHANGE_, _CONSENSUS BREAKING_, _build_, _chore_, _ci_, _docs_,_perf_, _refactor_, _revert_, _style_, _test_
    - `area`: _api_, _chain_, _state_, _vm_, _data transfer_, _market_, _mempool_, _message_, _block production_, _multisig_, _networking_, _paychan_, _proving_, _sealing_, _wallet_, _deps_
- [ ] This PR has tests for new functionality or change in behaviour
- [ ] If new user-facing features are introduced, clear usage guidelines and / or documentation updates should be included in https://lotus.filecoin.io or [Discussion Tutorials.](https://github.com/filecoin-project/lotus/discussions/categories/tutorials)
- [ ] CI is green

### CODE REVIEW

File: api/api_full.go [hunk 1]
```diff
@@ -522,8 +522,10 @@ type FullNode interface {
 	StateMarketStorageDeal(context.Context, abi.DealID, types.TipSetKey) (*MarketDeal, error) //perm:read
 	// StateLookupID retrieves the ID address of the given address
 	StateLookupID(context.Context, address.Address, types.TipSetKey) (address.Address, error) //perm:read
-	// StateAccountKey returns the public key address of the given ID address
+	// StateAccountKey returns the public key address of the given ID address for secp and bls accounts
 	StateAccountKey(context.Context, address.Address, types.TipSetKey) (address.Address, error) //perm:read
+	// StateLookupRobustAddress returns the public key address of the given ID address for non-account addresses (multisig, miners etc)
+	StateLookupRobustAddress(context.Context, address.Address, types.TipSetKey) (address.Address, error) //perm:read
```
Comment (Kubuxu) [inline api/api_full.go:528]: Looking at the implementation this for sure is slow due to data structures. Do we want to allow anyone (including for example browser) to call this?

File: chain/stmgr/stmgr.go [hunk 2]
```diff
@@ -318,6 +321,48 @@ func (sm *StateManager) LookupID(ctx context.Context, addr address.Address, ts *
 	return state.LookupID(addr)
 }
 
+func (sm *StateManager) LookupRobustAddress(ctx context.Context, idAddr address.Address, ts *types.TipSet) (address.Address, error) {
+	idAddrDecoded, err := address.IDFromAddress(idAddr)
+	if err != nil {
+		return address.Undef, xerrors.Errorf("address couldnt decode to id addr: %w", err)
```
Comment (arajasek) [inline chain/stmgr/stmgr.go:327]:
```suggestion
		return address.Undef, xerrors.Errorf("failed to decode provided address as id addr: %w", err)
```

File: node/impl/full/state.go [hunk 3]
```diff
@@ -449,6 +450,21 @@ func (m *StateModule) StateLookupID(ctx context.Context, addr address.Address, t
 	return m.StateManager.LookupID(ctx, addr, ts)
 }
 
+func (m *StateModule) StateLookupRobustAddress(ctx context.Context, addr address.Address, tsk types.TipSetKey) (address.Address, error) {
+	ts, err := m.Chain.GetTipSetFromKey(ctx, tsk)
+	if err != nil {
+		return address.Undef, xerrors.Errorf("loading tipset %s: %w", tsk, err)
+	}
+	if ts.Height() > policy.ChainFinality {
```
Comment (arajasek) [inline node/impl/full/state.go:458]: So, thinking about this some more, we aren't actually using a finality lookback when resolving `Account`s in `StateAccountKey`. While I think we probably should be, this method should just match behaviour there, so I'd actually drop the finality lookback here (in contrast to what I said we should do in the issue).

File: itests/self_sent_txn_test.go [hunk 4]
```diff
@@ -102,3 +102,23 @@ func TestSelfSentTxnV14(t *testing.T) {
 	require.NoError(t, err)
 	require.Equal(t, exitcode.Ok, mLookup.Receipt.ExitCode)
 }
+
+func TestStateLookupRobustAddress(t *testing.T) {
+	ctx := context.Background()
+	kit.QuietMiningLogs()
+
+	client15, miner, ens := kit.EnsembleMinimal(t, kit.MockProofs(), kit.GenesisNetworkVersion(network.Version15))
```
Comment (arajasek) [inline itests/self_sent_txn_test.go:110]:
```suggestion
	client, miner, ens := kit.EnsembleMinimal(t, kit.MockProofs(), kit.GenesisNetworkVersion(network.Version15))
```

Rename this to just `client`? `client15` was relevant to the other test in this file (it was relevant to network v15).

File: itests/self_sent_txn_test.go [hunk 5]
```diff
@@ -102,3 +102,23 @@ func TestSelfSentTxnV14(t *testing.T) {
 	require.NoError(t, err)
 	require.Equal(t, exitcode.Ok, mLookup.Receipt.ExitCode)
 }
+
+func TestStateLookupRobustAddress(t *testing.T) {
```
Comment (arajasek) [inline itests/self_sent_txn_test.go:106]: This test should move into its own file.

File: api/api_gateway.go [hunk 6]
```diff
@@ -54,6 +54,7 @@ type Gateway interface {
 	StateReadState(ctx context.Context, actor address.Address, tsk types.TipSetKey) (*ActorState, error) //perm:read
 	StateListMiners(ctx context.Context, tsk types.TipSetKey) ([]address.Address, error)
 	StateLookupID(ctx context.Context, addr address.Address, tsk types.TipSetKey) (address.Address, error)
+	StateLookupRobustAddress(ctx context.Context, addr address.Address, tsk types.TipSetKey) (address.Address, error)
```
Comment (arajasek) [inline api/api_gateway.go:57]: We definitely don't want this in the gateway (sloooow).

File: chain/stmgr/stmgr.go [hunk 7]
```diff
@@ -318,6 +321,48 @@ func (sm *StateManager) LookupID(ctx context.Context, addr address.Address, ts *
...
+	err = initState.ForEachActor(func(id abi.ActorID, addr address.Address) error {
+		if uint64(id) == idAddrDecoded {
+			robustAddr = addr
+			// Hacky way to early return from ForEach
+			return xerrors.New("robust address found")
+		}
+		return nil
+	})
+	if err != nil {
+		if robustAddr == address.Undef {
```
Comment (arajasek) [inline chain/stmgr/stmgr.go:358]: I would move this check to _before_ the error check -- assume that if robustAddr got set, you don't have an error (in contrast to actual correct error handling).

File: api/api_full.go (hunk 1 above)
Comment (Kubuxu) [reply api/api_full.go:528]: It would be best to bump the StateCompute perm as well (we could add another permission `slow` or somethign).

Comment (arajasek) [reply api/api_full.go:528]: Yeah, that's a good idea, but I think it'll be a future thing.

File: chain/stmgr/stmgr.go [hunk 8]
```diff
@@ -318,6 +321,48 @@ func (sm *StateManager) LookupID(ctx context.Context, addr address.Address, ts *
...
+		if uint64(id) == idAddrDecoded {
+			robustAddr = addr
+			// Hacky way to early return from ForEach
+			return xerrors.New("robust address found")
+		}
+		return nil
+	})
+	if robustAddr == address.Undef {
+		if err == nil {
+			return address.Undef, xerrors.Errorf("Address %s not found", idAddr.String())
```
Comment (arajasek) [inline chain/stmgr/stmgr.go:359]:
```suggestion
			return address.Undef, xerrors.Errorf("Address %d not found", idAddr)
```

**Review by arajasek — APPROVED** [review]

**Pull request merged**